File: src/resolver/resolve.ts
import * as path from "node:path";

import {
  ConflictError,
  CycleError,
  MissingIncludeError,
  MissingProfileError,
} from "../errors/index.js";

import { loadManifest } from "./manifest.js";
import { isMergeable, policyFor } from "./merge-policy.js";
import {
  buildPaths,
  classifyInclude,
  isValidProfileName,
  profileDir,
  type ResolverPaths,
} from "./paths.js";
import type {
  Contributor,
  ExternalTrustEntry,
  IncludeRef,
  PlanFile,
  ProfileManifest,
  ResolutionWarning,
  ResolvedPlan,
} from "./types.js";
import { RESOLVED_PLAN_SCHEMA_VERSION } from "./types.js";
import { isDirectory, walkClaudeDir, walkProfileRoot } from "./walk.js";

export interface ResolveOptions {
  /** Project root containing `.claude-profiles/`. */
  projectRoot: string;
}

/**
 * Resolve `profileName` into a ResolvedPlan.
 *
 * Algorithm (deterministic, single pass per phase):
 *  1. Walk extends chain upward, detecting cycles and missing profiles.
 *  2. Reverse to oldest-first order (R3 / R9).
 *  3. For each profile in chain, classify and validate its includes (R6/R7/R37).
 *     Includes contribute *after* their referencing profile's ancestors but
 *     before the next descendant. Per R9 worked example, the canonical order
 *     for `base ← extended ← profile` with `profile.includes = [A, B]` is:
 *     base, extended, A, B, profile.  So includes attach to whichever profile
 *     declared them, immediately AFTER that profile in the chain — but only
 *     for the *leaf* profile per the worked example. Other ancestors' includes
 *     come before them. We implement: walk oldest→newest; for each profile
 *     emit the profile, then its includes. EXCEPT the leaf — emit its
 *     includes *before* itself, so the leaf is last. This matches the worked
 *     example exactly.
 *  4. Walk every contributor's `.claude/` directory, collecting PlanFiles.
 *  5. Detect R11 conflicts on non-mergeable files.
 *  6. Sort files lex by relPath, stable by contributorIndex.
 *
 * Throws ResolverError subclasses on R4/R5/R7/R11. Returns warnings for R36.
 */
export async function resolve(
  profileName: string,
  opts: ResolveOptions,
): Promise<ResolvedPlan> {
  const paths = buildPaths(opts.projectRoot);
  const warnings: ResolutionWarning[] = [];

  // 1. Build extends chain (newest → oldest), detecting cycle/missing.
  const newestFirst = await buildExtendsChain(profileName, paths);
  // 2. Reverse to oldest-first (canonical order for R3, R9).
  const oldestFirst = [...newestFirst].reverse();

  // 3. Build contributors in canonical order. Per the R9 worked example,
  // includes for the leaf profile precede the leaf itself; ancestors' includes
  // are emitted immediately after the ancestor that declared them.
  const contributors: Contributor[] = [];
  const includes: IncludeRef[] = [];
  const externalPaths: ExternalTrustEntry[] = [];
  const seenExternal = new Set<string>();
  // Dedup include contributors by resolved path. Two raw strings (e.g.
  // "compA" and "./../_components/compA") that resolve to the same dir
  // would otherwise produce two contributors and falsely trip R11
  // self-conflicts on non-mergeable files. Each duplicate yields a warning.
  const seenContributorPaths = new Set<string>();

  const leafIndex = oldestFirst.length - 1;
  for (let i = 0; i < oldestFirst.length; i++) {
    const entry = oldestFirst[i]!;
    const isLeaf = i === leafIndex;

    if (!isLeaf) {
      // Ancestor: emit ancestor first, then its includes.
      contributors.push(makeAncestorContributor(entry.name, paths, entry.manifest));
      await emitIncludes(
        entry.includes,
        entry.dir,
        entry.name,
        paths,
        contributors,
        includes,
        externalPaths,
        seenExternal,
        seenContributorPaths,
        warnings,
      );
    } else {
      // Leaf: emit its includes first, then the leaf last.
      await emitIncludes(
        entry.includes,
        entry.dir,
        entry.name,
        paths,
        contributors,
        includes,
        externalPaths,
        seenExternal,
        seenContributorPaths,
        warnings,
      );
      contributors.push(makeProfileContributor(entry.name, paths, entry.manifest));
    }
  }

  // 4. Walk every contributor's .claude/ and collect files.
  const files: PlanFile[] = await collectFiles(contributors);

  // 5. Conflict detection (R11). Group files by the composite key
  // (relPath, destination). cw6/T3: a file at the same relPath in two
  // different destinations is, by construction, NOT a conflict — the merge
  // engine groups by the same composite key and produces two independent
  // MergedFile entries. R11 must therefore be scoped to a single destination
  // group as well; otherwise two contributors at e.g. `.claude/agents/x.json`
  // and `<projectRoot>/agents/x.json` (hypothetical) would falsely conflict.
  // Today this is benign because the only projectRoot file is CLAUDE.md
  // (concat policy → exempt from R11), but we encode the correct invariant
  // rather than relying on that policy coincidence.
  const byKey = new Map<string, PlanFile[]>();
  for (const f of files) {
    const key = `${f.destination}::${f.relPath}`;
    let arr = byKey.get(key);
    if (!arr) {
      arr = [];
      byKey.set(key, arr);
    }
    arr.push(f);
  }
  for (const group of byKey.values()) {
    if (group.length < 2) continue;
    const relPath = group[0]!.relPath;
    if (isMergeable(relPath)) continue; // R8/R9/R12 handle these in E2
    detectConflict(relPath, group, contributors);
  }

  // 6. Sort files: lex by relPath, then contributorIndex, then destination
  // (preserves canonical last-wins ordering when downstream consumers iterate;
  // destination tiebreak makes ordering deterministic when a single contributor
  // emits the same relPath to both destinations — cw6/T2).
  files.sort((a, b) => {
    if (a.relPath < b.relPath) return -1;
    if (a.relPath > b.relPath) return 1;
    if (a.contributorIndex !== b.contributorIndex)
      return a.contributorIndex - b.contributorIndex;
    if (a.destination === b.destination) return 0;
    return a.destination < b.destination ? -1 : 1;
  });

  // Aggregate manifest warnings.
  for (const entry of oldestFirst) {
    warnings.push(...entry.warnings);
  }

  const chain = oldestFirst.map((e) => e.name);

  return {
    schemaVersion: RESOLVED_PLAN_SCHEMA_VERSION,
    profileName,
    chain,
    includes,
    contributors,
    files,
    warnings,
    externalPaths,
  };
}

interface ChainEntry {
  /** Canonical profile name (directory name). */
  name: string;
  /** Absolute profile directory. */
  dir: string;
  /** Full parsed manifest (used for Contributor.manifest). */
  manifest: ProfileManifest;
  /** Includes from this profile's manifest, in declaration order. */
  includes: string[];
  /** Manifest-level warnings from loading this profile. */
  warnings: ResolutionWarning[];
}

async function buildExtendsChain(
  profileName: string,
  paths: ResolverPaths,
): Promise<ChainEntry[]> {
  const chain: ChainEntry[] = [];
  const visitedOrder: string[] = []; // for ordered cycle messages
  const visitedSet = new Set<string>(); // O(1) membership for cycle check
  let current: string | undefined = profileName;
  let referencedBy: string | undefined;

  while (current !== undefined) {
    // R2: profile identifiers are bare directory names. Reject anything
    // that could escape `.claude-profiles/` via traversal, slashes, or the
    // hidden/`_components` conventions before touching the filesystem.
    if (!isValidProfileName(current)) {
      throw new MissingProfileError(current, referencedBy);
    }

    if (visitedSet.has(current)) {
      // Cycle: slice starting at the first occurrence and append `current`
      // again so the cycle reads naturally: a → b → a.
      const start = visitedOrder.indexOf(current);
      const cycle = [...visitedOrder.slice(start), current];
      throw new CycleError(cycle);
    }
    visitedSet.add(current);
    visitedOrder.push(current);

    const dir = profileDir(paths, current);
    if (!(await isDirectory(dir))) {
      throw new MissingProfileError(current, referencedBy);
    }

    const { manifest, warnings: manifestWarnings } = await loadManifest(dir, current);

    // Track the leaf-side warnings on the entry; the resolver concatenates
    // them into ResolvedPlan.warnings later.
    chain.push({
      name: current,
      dir,
      manifest,
      includes: manifest.includes ?? [],
      warnings: manifestWarnings,
    });

    referencedBy = current;
    current = manifest.extends;
  }

  return chain;
}

async function emitIncludes(
  rawIncludes: string[],
  referencingProfileDir: string,
  referencedBy: string,
  paths: ResolverPaths,
  contributors: Contributor[],
  includes: IncludeRef[],
  externalPaths: ExternalTrustEntry[],
  seenExternal: Set<string>,
  seenContributorPaths: Set<string>,
  warnings: ResolutionWarning[],
): Promise<void> {
  for (const raw of rawIncludes) {
    const ref = classifyInclude(raw, referencingProfileDir, paths, referencedBy);
    includes.push(ref);

    if (!(await isDirectory(ref.resolvedPath))) {
      throw new MissingIncludeError(raw, ref.resolvedPath, referencedBy);
    }

    if (seenContributorPaths.has(ref.resolvedPath)) {
      // Same directory already contributed (possibly via a different raw
      // form). Skip to avoid duplicate Contributor entries; otherwise R11
      // would self-conflict on non-mergeable files and mergeable files
      // would be processed twice.
      warnings.push({
        code: "DuplicateInclude",
        message: `Include "${raw}" in profile "${referencedBy}" resolves to "${ref.resolvedPath}", which was already included; skipping duplicate`,
        source: referencedBy,
      });
      continue;
    }
    seenContributorPaths.add(ref.resolvedPath);

    contributors.push({
      kind: "include",
      id: raw,
      rootPath: ref.resolvedPath,
      claudeDir: path.join(ref.resolvedPath, ".claude"),
      external: ref.external,
    });

    if (ref.external && !seenExternal.has(ref.resolvedPath)) {
      seenExternal.add(ref.resolvedPath);
      externalPaths.push({ raw, resolvedPath: ref.resolvedPath });
    }
  }
}

function makeAncestorContributor(
  name: string,
  paths: ResolverPaths,
  manifest: ProfileManifest,
): Contributor {
  const dir = profileDir(paths, name);
  return {
    kind: "ancestor",
    id: name,
    rootPath: dir,
    claudeDir: path.join(dir, ".claude"),
    external: false,
    manifest,
  };
}

function makeProfileContributor(
  name: string,
  paths: ResolverPaths,
  manifest: ProfileManifest,
): Contributor {
  const dir = profileDir(paths, name);
  return {
    kind: "profile",
    id: name,
    rootPath: dir,
    claudeDir: path.join(dir, ".claude"),
    external: false,
    manifest,
  };
}

async function collectFiles(contributors: Contributor[]): Promise<PlanFile[]> {
  const out: PlanFile[] = [];
  for (let i = 0; i < contributors.length; i++) {
    const c = contributors[i]!;

    // 1. The historical .claude/ subtree → destination='.claude'.
    const claudeEntries = await walkClaudeDir(c.claudeDir);
    for (const e of claudeEntries) {
      out.push({
        relPath: e.relPath,
        absPath: e.absPath,
        contributorIndex: i,
        mergePolicy: policyFor(e.relPath),
        destination: ".claude",
      });
    }

    // 2. cw6/T2: profile-root files (CLAUDE.md peer-of-profile.json) →
    // destination='projectRoot'. Per spec §12 the merge policy is the same
    // (CLAUDE.md → 'concat'); only the destination grouping differs.
    const rootEntries = await walkProfileRoot(c.rootPath);
    for (const e of rootEntries) {
      out.push({
        relPath: e.relPath,
        absPath: e.absPath,
        contributorIndex: i,
        mergePolicy: policyFor(e.relPath),
        destination: "projectRoot",
      });
    }
  }
  return out;
}

/**
 * R11 conflict detection. Given multiple non-mergeable contributions for the
 * same relPath, decide if it's a conflict.
 *
 *  - profile-itself contributing → never a conflict (profile always overrides)
 *  - 2+ ancestors only           → no conflict (R10 last-wins among ancestors)
 *  - any include involved        → conflict
 */
function detectConflict(
  relPath: string,
  group: PlanFile[],
  contributors: Contributor[],
): void {
  const kinds = group.map((f) => contributors[f.contributorIndex]!.kind);
  if (kinds.includes("profile")) return; // profile overrides
  const hasInclude = kinds.includes("include");
  if (!hasInclude) return; // ancestor-only chain → R10 last-wins

  const offenders = group.map((f) => contributors[f.contributorIndex]!.id);
  // Deduplicate while preserving order (a single include directory could
  // theoretically be listed twice via different raw strings).
  const dedup = Array.from(new Set(offenders));
  throw new ConflictError(relPath, dedup);
}

File: internal/resolver/resolve.go
package resolver

import (
	"fmt"
	"path/filepath"
	"sort"

	pipelineerrors "github.com/htxryan/claude-code-config-profiles/internal/errors"
)

// ResolveOptions configures a Resolve call.
type ResolveOptions struct {
	ProjectRoot string
}

// Resolve resolves profileName into a ResolvedPlan. See the package doc and
// src/resolver/resolve.ts for the algorithm. Errors returned are pipeline
// error sub-types: *MissingProfileError, *CycleError, *MissingIncludeError,
// *ConflictError, *InvalidManifestError, *PathTraversalError.
func Resolve(profileName string, opts ResolveOptions) (*ResolvedPlan, error) {
	paths := BuildPaths(opts.ProjectRoot)
	warnings := []ResolutionWarning{}

	// 1. Build extends chain newest → oldest, detecting cycle/missing.
	newestFirst, err := buildExtendsChain(profileName, paths)
	if err != nil {
		return nil, err
	}

	// 2. Reverse to oldest-first.
	oldestFirst := make([]chainEntry, len(newestFirst))
	for i, e := range newestFirst {
		oldestFirst[len(newestFirst)-1-i] = e
	}

	// 3. Build contributors in canonical order.
	contributors := []Contributor{}
	includes := []IncludeRef{}
	externalPaths := []ExternalTrustEntry{}
	seenExternal := map[string]struct{}{}
	seenContributorPaths := map[string]struct{}{}
	leafIndex := len(oldestFirst) - 1
	for i, entry := range oldestFirst {
		isLeaf := i == leafIndex
		if !isLeaf {
			contributors = append(contributors, makeAncestorContributor(entry.name, paths, entry.manifest))
			if err := emitIncludes(entry.includes, entry.dir, entry.name, paths,
				&contributors, &includes, &externalPaths, seenExternal, seenContributorPaths, &warnings); err != nil {
				return nil, err
			}
		} else {
			if err := emitIncludes(entry.includes, entry.dir, entry.name, paths,
				&contributors, &includes, &externalPaths, seenExternal, seenContributorPaths, &warnings); err != nil {
				return nil, err
			}
			contributors = append(contributors, makeProfileContributor(entry.name, paths, entry.manifest))
		}
	}

	// 4. Walk every contributor's `.claude/` and root, collect files.
	files, err := collectFiles(contributors)
	if err != nil {
		return nil, err
	}
	if files == nil {
		files = []PlanFile{}
	}

	// 5. Conflict detection (R11). Group by (destination, relPath).
	byKey := map[string][]int{}
	for i, f := range files {
		key := string(f.Destination) + "::" + f.RelPath
		byKey[key] = append(byKey[key], i)
	}
	for _, idxs := range byKey {
		if len(idxs) < 2 {
			continue
		}
		relPath := files[idxs[0]].RelPath
		if IsMergeable(relPath) {
			continue
		}
		group := make([]PlanFile, 0, len(idxs))
		for _, idx := range idxs {
			group = append(group, files[idx])
		}
		if err := detectConflict(relPath, group, contributors); err != nil {
			return nil, err
		}
	}

	// 6. Sort files: lex by relPath, then contributorIndex, then destination.
	sort.SliceStable(files, func(i, j int) bool {
		if files[i].RelPath != files[j].RelPath {
			return files[i].RelPath < files[j].RelPath
		}
		if files[i].ContributorIndex != files[j].ContributorIndex {
			return files[i].ContributorIndex < files[j].ContributorIndex
		}
		return files[i].Destination < files[j].Destination
	})

	// Aggregate manifest warnings.
	for _, e := range oldestFirst {
		warnings = append(warnings, e.warnings...)
	}

	chain := make([]string, len(oldestFirst))
	for i, e := range oldestFirst {
		chain[i] = e.name
	}

	return &ResolvedPlan{
		SchemaVersion: ResolvedPlanSchemaVersion,
		ProfileName:   profileName,
		Chain:         chain,
		Includes:      includes,
		Contributors:  contributors,
		Files:         files,
		Warnings:      warnings,
		ExternalPaths: externalPaths,
	}, nil
}

type chainEntry struct {
	name     string
	dir      string
	manifest ProfileManifest
	includes []string
	warnings []ResolutionWarning
}

func buildExtendsChain(profileName string, paths ResolverPaths) ([]chainEntry, error) {
	var (
		chain        []chainEntry
		visitedOrder []string
		visitedSet   = map[string]struct{}{}
		current      = profileName
		referencedBy string
	)

	for {
		// R2: profile identifiers are bare directory names. Reject anything
		// that could escape `.claude-profiles/` via traversal, slashes, or
		// the hidden/`_components` conventions before touching the filesystem.
		if !IsValidProfileName(current) {
			return nil, pipelineerrors.NewMissingProfileError(current, referencedBy, nil)
		}

		if _, seen := visitedSet[current]; seen {
			start := indexOf(visitedOrder, current)
			cycle := append([]string{}, visitedOrder[start:]...)
			cycle = append(cycle, current)
			return nil, pipelineerrors.NewCycleError(cycle)
		}
		visitedSet[current] = struct{}{}
		visitedOrder = append(visitedOrder, current)

		dir := ProfileDir(paths, current)
		if !IsDirectory(dir) {
			return nil, pipelineerrors.NewMissingProfileError(current, referencedBy, nil)
		}

		res, err := LoadManifest(dir, current)
		if err != nil {
			return nil, err
		}

		chain = append(chain, chainEntry{
			name:     current,
			dir:      dir,
			manifest: res.Manifest,
			includes: append([]string{}, res.Manifest.Includes...),
			warnings: res.Warnings,
		})

		if res.Manifest.Extends == "" {
			return chain, nil
		}
		referencedBy = current
		current = res.Manifest.Extends
	}
}

func emitIncludes(
	rawIncludes []string,
	referencingProfileDir, referencedBy string,
	paths ResolverPaths,
	contributors *[]Contributor,
	includes *[]IncludeRef,
	externalPaths *[]ExternalTrustEntry,
	seenExternal, seenContributorPaths map[string]struct{},
	warnings *[]ResolutionWarning,
) error {
	for _, raw := range rawIncludes {
		ref, err := ClassifyInclude(raw, referencingProfileDir, paths, referencedBy)
		if err != nil {
			return err
		}
		// Append before the existence check so partial-progress shapes match
		// the TS resolver: on error, *includes is discarded by the caller
		// (Resolve returns nil, err). Do not switch this to a partial-plan
		// return without auditing every consumer for the new shape.
		*includes = append(*includes, ref)

		if !IsDirectory(ref.ResolvedPath) {
			return pipelineerrors.NewMissingIncludeError(raw, ref.ResolvedPath, referencedBy)
		}

		if _, dup := seenContributorPaths[ref.ResolvedPath]; dup {
			*warnings = append(*warnings, ResolutionWarning{
				Code:    WarningDuplicateInclude,
				Message: fmt.Sprintf("Include %q in profile %q resolves to %q, which was already included; skipping duplicate", raw, referencedBy, ref.ResolvedPath),
				Source:  referencedBy,
			})
			continue
		}
		seenContributorPaths[ref.ResolvedPath] = struct{}{}

		*contributors = append(*contributors, Contributor{
			Kind:      ContributorInclude,
			ID:        raw,
			RootPath:  ref.ResolvedPath,
			ClaudeDir: filepath.Join(ref.ResolvedPath, ".claude"),
			External:  ref.External,
		})

		if ref.External {
			if _, seen := seenExternal[ref.ResolvedPath]; !seen {
				seenExternal[ref.ResolvedPath] = struct{}{}
				*externalPaths = append(*externalPaths, ExternalTrustEntry{
					Raw:          raw,
					ResolvedPath: ref.ResolvedPath,
				})
			}
		}
	}
	return nil
}

func makeAncestorContributor(name string, paths ResolverPaths, manifest ProfileManifest) Contributor {
	dir := ProfileDir(paths, name)
	mc := manifest
	return Contributor{
		Kind:      ContributorAncestor,
		ID:        name,
		RootPath:  dir,
		ClaudeDir: filepath.Join(dir, ".claude"),
		External:  false,
		Manifest:  &mc,
	}
}

func makeProfileContributor(name string, paths ResolverPaths, manifest ProfileManifest) Contributor {
	dir := ProfileDir(paths, name)
	mc := manifest
	return Contributor{
		Kind:      ContributorProfile,
		ID:        name,
		RootPath:  dir,
		ClaudeDir: filepath.Join(dir, ".claude"),
		External:  false,
		Manifest:  &mc,
	}
}

func collectFiles(contributors []Contributor) ([]PlanFile, error) {
	var out []PlanFile
	for i, c := range contributors {
		// .claude/ subtree → destination='.claude'.
		entries, err := WalkClaudeDir(c.ClaudeDir)
		if err != nil {
			return nil, err
		}
		for _, e := range entries {
			out = append(out, PlanFile{
				RelPath:          e.RelPath,
				AbsPath:          e.AbsPath,
				ContributorIndex: i,
				MergePolicy:      PolicyFor(e.RelPath),
				Destination:      DestinationClaude,
			})
		}

		// profile-root files (CLAUDE.md peer of profile.json) → destination='projectRoot'.
		rootEntries, err := WalkProfileRoot(c.RootPath)
		if err != nil {
			return nil, err
		}
		for _, e := range rootEntries {
			out = append(out, PlanFile{
				RelPath:          e.RelPath,
				AbsPath:          e.AbsPath,
				ContributorIndex: i,
				MergePolicy:      PolicyFor(e.RelPath),
				Destination:      DestinationProjectRoot,
			})
		}
	}
	return out, nil
}

// detectConflict implements R11: given multiple non-mergeable contributions
// for the same (destination, relPath), decide if it's a conflict.
//   - profile-itself contributing → never a conflict (profile always overrides)
//   - 2+ ancestors only           → no conflict (R10 last-wins)
//   - any include involved        → conflict
func detectConflict(relPath string, group []PlanFile, contributors []Contributor) error {
	hasProfile, hasInclude := false, false
	for _, f := range group {
		k := contributors[f.ContributorIndex].Kind
		if k == ContributorProfile {
			hasProfile = true
		}
		if k == ContributorInclude {
			hasInclude = true
		}
	}
	if hasProfile {
		return nil
	}
	if !hasInclude {
		return nil
	}

	seen := map[string]struct{}{}
	dedup := []string{}
	for _, f := range group {
		id := contributors[f.ContributorIndex].ID
		if _, dup := seen[id]; dup {
			continue
		}
		seen[id] = struct{}{}
		dedup = append(dedup, id)
	}
	return pipelineerrors.NewConflictError(relPath, dedup)
}

func indexOf(s []string, v string) int {
	for i, x := range s {
		if x == v {
			return i
		}
	}
	return -1
}
